Ctop improvements

balajivenki opened the pull request:
@AntonGabov and @chemist777 As we discussed I have modified the library to boot from a properties file and also added the capability to publish the metrics to Graphite. Please review the code changes and let me know if anything needs to be modified. I appreciate your time on reviewing this PR in advance.

balajivenki committed:
added jmx username and password support

balajivenki committed:
code cleanup

balajivenki committed:
Added support for loading from properties file and also a ALL metrics mode which prints metrics for all keyspaces than the given one

balajivenki committed:
Added graphite reporter and RegEx mode

balajivenki committed:
Added graphite metrics

balajivenki committed:
Code cleanup

balajivenki committed:
Code cleanup

balajivenki committed:
Modified Readme

balajivenki committed:
Modified Readme

balajivenki committed:
Modified Readme

balajivenki committed:
Modified Readme

balajivenki committed:
Updated README.md

balajivenki commented:
@AntonGabov and @chemist777  any update on this PR?

AntonGabov commented on pom.xml:
```diff
@@ -13,6 +13,41 @@
+        </dependency>
+
+        <!-- https://mvnrepository.com/artifact/org.projectlombok/lombok -->
+        <dependency>
```
I don't appreciate of using lombock because developer has to install additional plugin in IDE

balajivenki replied:
Good point. Removed lombok dependency

AntonGabov commented on pom.xml:
```diff
@@ -13,6 +13,41 @@
+        <dependency>
+            <groupId>com.ryantenney.metrics</groupId>
+            <artifactId>metrics-spring</artifactId>
+            <version>3.1.2</version>
```
You can define a property which is related to metrics version like:
 <properties>
    <dropwizard.metrics>3.1.2</dropwizard.metrics>
</properties>

balajivenki replied:
Done

AntonGabov commented on src/main/java/com/pixonic/ctop/Main.java:
```diff
@@ -1,45 +1,105 @@
+
+        int interval = Integer.parseInt(properties.getProperty(Constants.CONFIG_INTERVAL_SEC, "10"));
+
+        String jmxUsername = properties.getProperty(Constants.CONFIG_JMX_USERNAME).trim();
```
Potential NPE because we can skip declaration of the variable 'jmx username' in properties file.

balajivenki replied:
Good point. Done

AntonGabov commented on src/main/java/com/pixonic/ctop/Main.java:
```diff
@@ -1,45 +1,105 @@
+        int interval = Integer.parseInt(properties.getProperty(Constants.CONFIG_INTERVAL_SEC, "10"));
+
+        String jmxUsername = properties.getProperty(Constants.CONFIG_JMX_USERNAME).trim();
+        String jmxPassword = properties.getProperty(Constants.CONFIG_JMX_PASSWORD).trim();
```
Potential NPE because we can skip declaration of this variable  in properties file.

balajivenki replied:
Done

AntonGabov commented on src/main/java/com/pixonic/ctop/Main.java:
```diff
@@ -1,45 +1,105 @@
+        System.out.println("======= Properties Loaded =======");
+        System.out.println(properties.toString());
+
+        MetricsType metricsType = MetricsType.valueOf(properties.getProperty("metrics.type", "NONE"));
```
I think it is not good idea to have "NONE" type because we will collect metrics and don't print it at all.
By default "CONSOLE" is enough.

balajivenki replied:
I will change it to CONSOLE, however NONE will enact the old behavior or printing in the console as formatted output. CONSOLE will print like normal graphite data in the console.

AntonGabov commented on src/main/java/com/pixonic/ctop/Main.java:
```diff
@@ -1,45 +1,105 @@
-        Metrics metrics = MetricsFactory.getMetrics(majorVersion, interval, remote, keySpace);
+        TargetType targetType = TargetType.valueOf(properties.getProperty(Constants.CONFIG_TARGET_TYPE, "ALL"));
+
+        Metrics metrics = MetricsFactory.getMetrics(majorVersion, interval, remote, keySpace, targetType, metricsType, metricsCollector);
```
metricsType is redundant argument

AntonGabov commented on src/main/java/com/pixonic/ctop/Main.java:
```diff
@@ -1,45 +1,105 @@
         metrics.printMetrics();
 
-        Runtime.getRuntime().addShutdownHook(new Thread(metrics::shutdown));
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> metrics.shutdown()));
```
'new Thread(metrics::shutdown)' looks pretty :)

AntonGabov commented on src/main/java/com/pixonic/ctop/Main.java:
```diff
@@ -1,45 +1,105 @@
+        //customer file passed so load it
+        if (args.length > 0) {
+            String argument = args[0].trim();
+            if (argument.equals("-h")) {
```
I don't like idea that we have to provide config.properties in order to have console metrics.
We should keep previous implementation of ctop for console.

You can implement 2 options:
1. Console option: use program arguments (like in initial implementation of ctop) or config.properties
2. Graphite option: use only config.properties

AntonGabov commented on src/main/java/com/pixonic/ctop/Main.java:
```diff
@@ -1,45 +1,105 @@
+                    System.out.println("File " + argument + " doesn't exist.");
+                    return;
+                } catch (Exception e) {
+                    e.printStackTrace();
```
Maybe we should also print this info to console?

AntonGabov commented on src/main/java/com/pixonic/ctop/Main.java:
```diff
@@ -1,45 +1,105 @@
+                System.out.println("File " + propertyFileName + " doesn't exist.");
+                return;
+            } catch (Exception e) {
+                e.printStackTrace();
```
Maybe we should also print this info to console?

AntonGabov commented on src/main/java/com/pixonic/ctop/metrics/AbstractMetrics.java:
```diff
@@ -14,11 +14,17 @@
+    protected final MetricsCollector metricsCollector;
 
-    AbstractMetrics(long interval, MBeanServerConnection remote, String keySpace) {
+    AbstractMetrics(long interval, MBeanServerConnection remote, String keySpace, TargetType targetType, MetricsType metricsType, MetricsCollector metricsCollector) {
```
metricsType is redundant argument

AntonGabov commented on src/main/java/com/pixonic/ctop/metrics/AbstractMetrics.java:
```diff
@@ -42,33 +48,66 @@ void printMetrics(NavigableSet<ResultItem> readResult, NavigableSet<ResultItem>
+
+        Long totalReadCount = 0L;
+        Long totalWriteCount = 0L;
+        boolean isMetricsEnabled = !metricsType.equals(MetricsType.NONE);
```
MetricsType.NONE should be removed. By default, CONSOLE can be used

AntonGabov commented on src/main/java/com/pixonic/ctop/metrics/AbstractMetrics.java:
```diff
@@ -42,33 +48,66 @@ void printMetrics(NavigableSet<ResultItem> readResult, NavigableSet<ResultItem>
-        return StringUtils.leftPad(String.valueOf(resultItem.count), maxLen + 1) + " " + resultItem;
+
+        //when the mode is not a single Keyspace then prefix the keyspace name to the metric to know from which keyspace it is
+        if (targetType.equals(TargetType.KEYSPACE)) {
```
Potential NPE.
To prevent NPE enums can be compared like:
1) ==
2) TargetType.KEYSPACE.equals(targetType)

balajivenki replied:
When TargetType(cassandra.target.type) is not given it is by default assumed as ALL(all keyspaces) and it is done in Main.java

`TargetType targetType = TargetType.valueOf(properties.getProperty(Constants.CONFIG_TARGET_TYPE, "ALL"));`

AntonGabov commented on src/main/java/com/pixonic/ctop/metrics/CurrentMetrics.java:
```diff
@@ -13,13 +13,15 @@
 
-    CurrentMetrics(long interval, MBeanServerConnection remote, String keySpace) {
-        super(interval, remote, keySpace);
+    CurrentMetrics(long interval, MBeanServerConnection remote, String keySpace, TargetType targetType, MetricsType metricsType, MetricsCollector metricsCollector) {
```
metricsType is redundant

AntonGabov commented on src/main/java/com/pixonic/ctop/metrics/CurrentMetrics.java:
```diff
@@ -13,13 +13,15 @@
-        ObjectName writeObjectName = new ObjectName("org.apache.cassandra.metrics:type=Table,keyspace=" + keySpace + ",scope=*,name=WriteLatency");
+    @Override
+    public void printMetrics() throws Exception {
+        String ksValue = targetType.equals(TargetType.ALL) ? "*" : keySpace;
```
Potential NPE

balajivenki replied:
Changed to 

`String ksValue = TargetType.ALL.equals(targetType) ? "*" : keySpace;`

AntonGabov commented on src/main/java/com/pixonic/ctop/metrics/OldMetrics.java:
```diff
@@ -11,14 +11,17 @@
+    @Override
+    public void printMetrics() throws Exception {
+        // when Mode is All then just use *, other types like Keyspace and RegEx can be passed as it is
+        String ksValue = targetType.equals(TargetType.ALL) ? "*" : keySpace;
```
Potential NPE

balajivenki replied:
Changed to 

`String ksValue = TargetType.ALL.equals(targetType) ? "*" : keySpace;`

AntonGabov commented on src/main/java/com/pixonic/ctop/util/MetricUtils.java:
```diff
@@ -0,0 +1,69 @@
+                break;
+            }
+
+            case NONE:
```
By default we can use CONSOLE instead of NONE

balajivenki replied:
NONE means the actual or legacy formatted output in console.

AntonGabov commented:
@balajivenki sorry for such late response.
I wrote some comments on your code. It's cool that you have implemented integration with Graphite!

balajivenki committed:
Code review comments and suggestions implemented

balajivenki commented:
Thanks @AntonGabov 

I have modified the code with your review suggestions. Please review it and let me know if I am missing anything still. Appreciate your time.

AntonGabov commented:
As I wrote in comments we would like to keep old functionality of running ctop without config.properties:
```
java -jar ctop.jar <host:jmx_port> <key_space>
```
So, I suggest restore old functionality (transfer program arguments to main function) and keep your code in order to send metrics to Graphite (use config.properties)

balajivenki commented:
@AntonGabov I will have to do changes for that though. When any even empty property file sent the default values and behavior fallback to what the old functionality was. I will accoyyour suggestion today and will update the pr.

balajivenki committed:
Code review suggestions and backward compatibility

balajivenki commented:
@AntonGabov I have added the backward compatibility as you recommended and also updated the README.md with the proper details. Please review the same and let me know your comments.

AntonGabov commented:
@balajivenki hello.
If we run ctop with program arguments (including default interval = 10 sec.) and "CONSOLE" type, every second the same info will be printed and only after 10 seconds it will be refreshed, but MetricRegistry will continue print old info each second.

From my point of view, type "CONSOLE" is redundant. Maybe I'm not right :)
Could you explain difference between "CONSOLE" and "NONE"?

balajivenki committed:
Not printing in console when metrics type is NONE

balajivenki commented:
@AntonGabov Yes CONSOLE printing prints the actual metrics in the format we send to graphite. Usually console metrics are used if some other application wants to read the data and port it to something or if we use any adapter for prometheus etc.

Both CONSOLE and GRAPHITE are accumulative data to plot the graph and so if this minute a cf has read 25 and next minute there is no read it will continue the data as 25 for the cf so the graph shows the accurate values.

NONE will merely mimic the actual behavior without the program arguments and so yes NONE and no program arguments are redundant and not CONSOLE and NONE.

I would prefer having NONE to distinguish and also have the way to print the formatted output using properties file as well. Hope this explains well.

chemist777 commented on README.md:
```diff
@@ -6,16 +6,78 @@ You can see which column families mostly affects disk utilization in near real t
 ## Usage
+- When using a properties file
+`java -jar ctop.jar -p [<propertyFilePath> when not passed ctop.properties will be loaded from classpath with argument -p]`
+- When not using a properties file
```
Please change the order. Example without property file should be first in the list.

chemist777 commented on README.md:
```diff
@@ -6,16 +6,78 @@ You can see which column families mostly affects disk utilization in near real t
+- When using a properties file
+`java -jar ctop.jar -p [<propertyFilePath> when not passed ctop.properties will be loaded from classpath with argument -p]`
+- When not using a properties file
+`java -jar ctop.jar <host:jmx_port> <key_space> [interval_sec(default: 10)] when doesn't want to use any properties file`
```
Move ending single quote to enclose command only.

chemist777 commented on README.md:
```diff
@@ -6,16 +6,78 @@ You can see which column families mostly affects disk utilization in near real t
+- When not using a properties file
+`java -jar ctop.jar <host:jmx_port> <key_space> [interval_sec(default: 10)] when doesn't want to use any properties file`
+- To print help usage
+`java -jar ctop.jar -h to print help`
```
Move ending single quote to enclose command only.

chemist777 commented on README.md:
```diff
@@ -6,16 +6,78 @@ You can see which column families mostly affects disk utilization in near real t
 
 ## Usage
+- When using a properties file
+`java -jar ctop.jar -p [<propertyFilePath> when not passed ctop.properties will be loaded from classpath with argument -p]`
```
Don't write comments in the single-line code block.

chemist777 reviewed (changes requested):
Please change doc.

balajivenki committed:
Code review comments on README.md

balajivenki commented:
Thanks @chemist777  for the comments. I have fixed the docs and committed it. Please review it.

balajivenki commented:
Thanks a lot @chemist777 and @AntonGabov
